torch/_custom_op/autograd.py

```diff
@@ -56,6 +56,37 @@ def kernel(*args, **kwargs):
     return kernel
 
 
+def mark_non_differentiable(ctx, output, output_differentiability):
+    # Output types are restricted to be:
```

Comment:
Aren't all optional variants of these ok?

Reply:
I didn't want to deal with them, plus no built-in operators actually return optional versions of these, so we are restricting it to this set to make the implementation easier. I can extend these to the optional variants if you think we should have that now?

Reply:
Sounds good to me!